editor/src/components/inspector/controls/color-picker-swatches.tsx

```diff
@@ -0,0 +1,185 @@
+      setEditing(false)
+    }
+  }, [colorSwatches])
+
+  const toggleEditing = React.useCallback(() => {
+    setEditing(!editing)
+  }, [editing])
+
+  const onAddPreset = React.useCallback(() => {
+    const preset: ColorSwatch = {
+      id: UUID(),
+      hex: currentColor,
+    }
+    setColorSwatches(colorSwatches.concat(preset))
+  }, [currentColor, colorSwatches])
+
+  const onClickPreset = React.useCallback(
+    (c: ColorSwatch) => () => {
+      if (editing) {
+        setColorSwatches(colorSwatches.filter((p) => p.id != c.id))
+      } else {
+        onSelectColor(c.hex)
+      }
+    },
+    [editing, onSelectColor, colorSwatches],
+  )
+
+  return (
+    <FlexColumn
+      style={{
+        padding: '0px 8px 4px 8px',
+        marginTop: 4,
+        gap: 4,
+      }}
+    >
+      <FlexRow style={{ justifyContent: 'space-between' }}>
+        <div style={{ fontWeight: 600 }}>Swatches</div>
+        {colorSwatches.length > 0 && (
+          <Button spotlight highlight style={{ padding: '0 6px' }} onMouseDown={toggleEditing}>
+            {editing ? 'Done' : 'Edit'}
+          </Button>
+        )}
+      </FlexRow>
+
+      <FlexRow style={{ flexWrap: 'wrap', gap: 2 }}>
+        {colorSwatches.map((c, index) => {
+          return (
+            <Button
+              onClick={onClickPreset(c)}
+              key={index}
+              title={editing ? 'Remove color' : ''}
+              style={{
+                position: 'relative',
+                width: SWATCH_SIZE,
+                height: SWATCH_SIZE,
+                background: c.hex,
+                border: `1px solid ${
+                  currentColor === c.hex ? colorTheme.fg0.value : colorTheme.border0.value
+                }`,
+                borderRadius: 2,
+                zIndex: 0,
+              }}
+            >
+              {editing && <TrashIcon />}
+              {!editing && currentColor === c.hex && <CurrentColorBadge />}
+            </Button>
+          )
+        })}
+
+        {!editing && (
+          <Button
+            spotlight
+            highlight
+            style={{
+              padding: '0 6px',
+              width: SWATCH_SIZE,
+              height: SWATCH_SIZE,
+            }}
+            onMouseDown={onAddPreset}
+          >
+            <PlusIcon />
+          </Button>
+        )}
+      </FlexRow>
+    </FlexColumn>
+  )
+})
+
+const CurrentColorBadge = React.memo(() => {
+  return (
+    <div
+      style={{
+        position: 'absolute',
+        top: -2,
+        right: -2,
+        width: 7,
+        height: 7,
+        background: colorTheme.primary.value,
+        border: `1px solid ${colorTheme.border0.value}`,
+        borderRadius: '100%',
+        zIndex: 1,
+        boxShadow: `${colorTheme.canvasControlsSizeBoxShadowColor20.value} 0px 0px 1px, ${colorTheme.canvasControlsSizeBoxShadowColor21.value} 0px 1px 2px 1px`,
+      }}
+    />
+  )
+})
+
+const TrashIcon: React.FC = () => {
```

Comment:
minor but you can memoize TrashIcon and PlusIcon too

Reply:
https://github.com/concrete-utopia/utopia/pull/3133/commits/e56f43d4ed54256d7e1cb8fc0051b0b7566fb9c1